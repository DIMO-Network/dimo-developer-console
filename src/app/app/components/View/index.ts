import View from './View';

export { View };

export default View;
