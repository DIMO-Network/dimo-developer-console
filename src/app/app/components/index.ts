export * from './OnboardingCard';
export * from './Onboarding';
