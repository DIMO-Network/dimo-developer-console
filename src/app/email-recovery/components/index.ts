export * from './EmailRecoveryForm';
export * from './CheckEmail';