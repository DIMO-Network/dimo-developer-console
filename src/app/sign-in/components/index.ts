export * from './OtpInputForm';
export * from './PasskeyLogin';
export * from './SignInMethodForm';
