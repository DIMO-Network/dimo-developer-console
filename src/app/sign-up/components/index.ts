export * from './BuildForForm';
export * from './CompanyInfoForm';
export * from './SignUpWith';
export * from './UserInfoForm';
