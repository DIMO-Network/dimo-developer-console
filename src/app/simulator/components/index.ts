export { View } from './View';
export { Header } from './Header';
export { SimulatedVehiclesSection } from './SimulatedVehiclesSection';
export { CreateSimulatedVehicleModal } from './CreateSimulatedVehicleModal';
