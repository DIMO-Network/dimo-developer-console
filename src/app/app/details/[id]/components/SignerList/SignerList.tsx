import _ from 'lodash';

import { maskStringV2 } from 'maskdata';
import { TrashIcon } from '@heroicons/react/24/outline';
import { useState, type FC } from 'react';
import { encodeFunctionData } from 'viem';

import { Button } from '@/components/Button';
import { ContentCopyIcon } from '@/components/Icons';
import { deleteMySigner, testApp } from '@/actions/app';
import { IApp, ISigner } from '@/types/app';
import { LoadingModal, LoadingProps } from '@/components/LoadingModal';
import { Table } from '@/components/Table';
import { TeamRoles } from '@/types/team';
import { useContractGA, useGlobalAccount, useOnboarding } from '@/hooks';
import { useSession } from 'next-auth/react';
import DimoLicenseABI from '@/contracts/DimoLicenseContract.json';

import configuration from '@/config';

interface IProps {
  app: IApp;
  refreshData: () => void;
}

export const SignerList: FC<IProps> = ({ app, refreshData }) => {
  const [isOpened, setIsOpened] = useState<boolean>(false);
  const [loadingStatus, setLoadingStatus] = useState<LoadingProps>();
  const { workspace } = useOnboarding();
  const { organizationInfo } = useGlobalAccount();
  const { processTransactions } = useContractGA();
  const { data: session } = useSession();
  const { user: { role = '' } = {} } = session ?? {};

  const handleCopy = (value: string) => {
    void navigator.clipboard.writeText(value);
  };

  const handleDisableSigner = async (signer: string) => {
    if (!organizationInfo && !workspace)
      throw new Error('Web3 connection failed');
    const transaction = [{
      to: configuration.DLC_ADDRESS,
      value: BigInt(0),
      data: encodeFunctionData({
        abi: DimoLicenseABI,
        functionName: 'disableSigner',
        args: [
          workspace?.token_id ?? 0,
          signer,
        ]
      }),
    }];
    await processTransactions(transaction);
  };

  const renderWithCopy = (columnName: string, data: Record<string, string>) => {
    const value = String(data[columnName]).replace('0x', '');
    return (
      <p className="flex flex-row">
        {maskStringV2(value, {
          maskWith: '*',
          unmaskedEndCharacters: 2,
          unmaskedStartCharacters: 2,
          maxMaskedCharacters: 20,
        })}
        <ContentCopyIcon
          className="w-4 h-4 ml-2 fill-white/50 cursor-pointer"
          onClick={() => handleCopy(value)}
        />
      </p>
    );
  };

  const handleTestAuthentication = async (signer: ISigner) => {
    try {
      setIsOpened(true);
      setLoadingStatus({
        label: 'Testing the application',
        status: 'loading',
      });
      const { uri: domain = '' } =
        app.RedirectUris?.find(({ deleted }) => !deleted) || {};
      if (!domain) {
        return setLoadingStatus({
          label: 'You need to set at least one domain',
          status: 'error',
        });
      }
      await testApp(app, signer);
      setLoadingStatus({
        label: 'Application tested successfully',
        status: 'success',
      });
    } catch (error: unknown) {
      setLoadingStatus({ label: 'Something went wrong', status: 'error' });
    }
  };

  const renderTestAuthenticationAction = (signer: ISigner) => {
    return (
      <Button
        className="white-outline px-4"
        onClick={() => handleTestAuthentication(signer)}
        key={signer.id}
      >
        Test Authentication
      </Button>
    );
  };

  const renderDeleteSignerAction = ({
    id = '',
    api_key: signer = '',
  }: ISigner) => {
    return (
      role === TeamRoles.OWNER && (
        <button type="button" onClick={() => handleDelete(id, signer)} key={id}>
          <TrashIcon className="w-5 h-5" />
        </button>
      )
    );
  };

  const handleDelete = async (id: string, signer: string) => {
    try {
      setIsOpened(true);
      await handleDisableSigner(signer);
      setLoadingStatus({
        label: 'Deleting the selected API key',
        status: 'loading',
      });
      await deleteMySigner(id);
      setLoadingStatus({ label: 'API key deleted', status: 'success' });
      refreshData();
    } catch (error: unknown) {
      const code = _.get(error, 'code', null);
      if (code === 4001)
        setLoadingStatus({
          label: 'The transaction was denied',
          status: 'error',
        });
      else setLoadingStatus({ label: 'Something went wrong', status: 'error' });
    }
  };

  return (
    <>
      {app.Signers && app.Signers.length > 0 && (
        <>
          <Table
            columns={[
              {
                name: 'api_key',
                label: 'API Key',
                render: (item) => renderWithCopy('api_key', item),
              },
            ]}
            data={app.Signers.filter(({ deleted }) => !deleted)}
            actions={[renderTestAuthenticationAction, renderDeleteSignerAction]}
          />
          <LoadingModal
            isOpen={isOpened}
            setIsOpen={setIsOpened}
            {...loadingStatus}
          />
        </>
      )}
    </>
  );
};

export default SignerList;
