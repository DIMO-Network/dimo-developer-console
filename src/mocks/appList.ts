import { IApp } from '@/types/app';

export const appListMock: IApp[] = [
  {
    name: 'Test App',
    scope: 'sandbox',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
  {
    name: 'Fleet Management App',
    scope: 'production',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
  {
    name: 'Data Analytics',
    scope: 'production',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
  {
    name: 'Hello World',
    scope: 'production',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
  {
    name: 'Hello World',
    scope: 'production',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
  {
    name: 'Hello World',
    scope: 'production',
    client_id: '0x*HSD)BN*(SHJDM)S(IJD SIOPDJSN)abcded1EA',
    signers: [
      {
        wallet: '0x1ABC7154748D1CE5144478CDEB574AE244B939B5',
        key: 'abcdef12345abcdef12345abcdef12345abcdef12345',
      },
    ],
    redirectUris: [
      {
        redirectUri: 'https://google.com/',
        status: 'active',
      },
      {
        redirectUri: 'https://google.com.co/',
        status: 'inactive',
      },
    ],
  },
];
