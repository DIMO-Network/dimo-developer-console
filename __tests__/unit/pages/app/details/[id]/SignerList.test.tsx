import { fireEvent, render, screen } from '@testing-library/react';

import { SignerList } from '@/app/app/details/[id]/components/SignerList';
import { appListMock } from '@/mocks/appList';

const [app] = appListMock;

const writeTextTest = jest.fn();
Object.assign(navigator, {
  clipboard: {
    writeText: writeTextTest,
  },
});

describe('SignerList', () => {
  it('renders the app summary', () => {
    const { container } = render(
      <SignerList list={app.Signers} refreshData={() => {}} />,
    );

    const [signer] = app.Signers ?? [];
    const [keyElm] = screen.getAllByText(signer.api_key);
    const [contentCopyElm] = screen.getAllByRole('content-copy-icon');

    fireEvent.click(contentCopyElm);

    expect(keyElm).toBeInTheDocument();
    expect(contentCopyElm).toBeInTheDocument();
    expect(writeTextTest).toHaveBeenCalledWith(signer.api_key);
    expect(container).toMatchSnapshot();
  });
});
