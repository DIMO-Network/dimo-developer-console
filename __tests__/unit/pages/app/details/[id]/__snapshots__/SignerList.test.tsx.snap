// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignerList renders the app summary 1`] = `
<div>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          class="custom-table-column"
          scope="col"
        >
          wallet
        </th>
        <th
          class="custom-table-column"
          scope="col"
        >
          key
        </th>
        <th
          class="table-action-column"
          scope="col"
        >
          <span
            class="sr-only"
          >
            Actions
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      class="table-body"
    >
      <tr>
        <td
          class="table-cell"
        >
          0x1ABC7154748D1CE5144478CDEB574AE244B939B5
        </td>
        <td
          class="table-cell"
        >
          <p
            class="flex flex-row"
          >
            abcdef12345abcdef12345abcdef12345abcdef12345
            <svg
              class="w-4 h-4 ml-2 fill-white/50 cursor-pointer"
              fill="none"
              role="content-copy-icon"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7497 0.833374H3.74967C2.83301 0.833374 2.08301 1.58337 2.08301 2.50004V14.1667H3.74967V2.50004H13.7497V0.833374ZM16.2497 4.16671H7.08301C6.16634 4.16671 5.41634 4.91671 5.41634 5.83337V17.5C5.41634 18.4167 6.16634 19.1667 7.08301 19.1667H16.2497C17.1663 19.1667 17.9163 18.4167 17.9163 17.5V5.83337C17.9163 4.91671 17.1663 4.16671 16.2497 4.16671ZM16.2497 17.5H7.08301V5.83337H16.2497V17.5Z"
              />
            </svg>
          </p>
        </td>
        <td
          class="table-action-cell"
        >
          <button
            class="button white-outline px-4"
          >
            <span>
              Test Authentication
            </span>
          </button>
          <svg
            aria-hidden="true"
            class="w-5 h-5"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </td>
      </tr>
      <tr>
        <td
          class="table-cell"
        >
          0x1ABC7154748D1CE5144478CDEB574AE244B939B5
        </td>
        <td
          class="table-cell"
        >
          <p
            class="flex flex-row"
          >
            abcdef12345abcdef12345abcdef12345abcdef12345
            <svg
              class="w-4 h-4 ml-2 fill-white/50 cursor-pointer"
              fill="none"
              role="content-copy-icon"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7497 0.833374H3.74967C2.83301 0.833374 2.08301 1.58337 2.08301 2.50004V14.1667H3.74967V2.50004H13.7497V0.833374ZM16.2497 4.16671H7.08301C6.16634 4.16671 5.41634 4.91671 5.41634 5.83337V17.5C5.41634 18.4167 6.16634 19.1667 7.08301 19.1667H16.2497C17.1663 19.1667 17.9163 18.4167 17.9163 17.5V5.83337C17.9163 4.91671 17.1663 4.16671 16.2497 4.16671ZM16.2497 17.5H7.08301V5.83337H16.2497V17.5Z"
              />
            </svg>
          </p>
        </td>
        <td
          class="table-action-cell"
        >
          <button
            class="button white-outline px-4"
          >
            <span>
              Test Authentication
            </span>
          </button>
          <svg
            aria-hidden="true"
            class="w-5 h-5"
            data-slot="icon"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m14.74 9-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 0 1-2.244 2.077H8.084a2.25 2.25 0 0 1-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 0 0-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 0 1 3.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 0 0-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 0 0-7.5 0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
